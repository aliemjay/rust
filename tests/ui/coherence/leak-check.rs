// compile-flags: -Ztrait-solver=next
trait Relate {}

struct Outlives<'a, 'b>(&'a u8, &'b u8);
impl<'a, 'b> Relate for Outlives<'a, 'b> where 'a: 'b, {}

struct Equal<'a, 'b>(&'a u8, &'b u8);
impl<'a> Relate for Equal<'a, 'a> {}

macro_rules! rule {
    ( $name:ident<$($lt:lifetime),*> :- $( ($($bound:tt)*) ),* ) => {
        struct $name<$($lt),*>($(&$lt u8),*);
        impl<$($lt),*> $crate::Relate for $name<$($lt),*>
            where $( $($bound)*: $crate::Relate, )*
        {}
    };
}

// ----

trait CoherenceTrait {}
impl<T> CoherenceTrait for T {}

macro_rules! assert_false_by_leak_check {
    ( exist<$($lt:lifetime),*> $( ($($bound:tt)*) ),* ) => {
        struct Unique<$($lt),*>($(&$lt u8),*);
        impl<$($lt),*> $crate::CoherenceTrait for Unique<$($lt),*>
            //~^ ERROR for type `test1::Unique`
            //~| ERROR for type `test3::Unique`
            //~| ERROR for type `test6::Unique`
            where $( $($bound)*: $crate::Relate, )*
        {}
    };
}

mod test1 {
    use super::*;
    assert_false_by_leak_check!(
        exist<> (for<'a, 'b> Outlives<'a, 'b>)
    );
}

mod test2 {
    use super::*;
    assert_false_by_leak_check!(
        exist<'a> (for<'b> Outlives<'b, 'a>)
    );
}

mod test3 {
    use super::*;
    rule!( OutlivesPlaceholder<'a> :- (for<'b> Outlives<'a, 'b>) );
    assert_false_by_leak_check!(
        exist<> (for<'a> OutlivesPlaceholder<'a>)
    );
}

mod test4 {
    use super::*;
    rule!( OutlivedByPlaceholder<'a> :- (for<'b> Outlives<'b, 'a>) );
    assert_false_by_leak_check!(
        exist<> (for<'a> OutlivedByPlaceholder<'a>)
    );
}

mod test5 {
    use super::*;
    rule!( EqualsPlaceholder<'a> :- (for<'b> Equal<'b, 'a>) );
    assert_false_by_leak_check!(
        exist<> (for<'a> EqualsPlaceholder<'a>)
    );
}

mod test6 {
    use super::*;
    assert_false_by_leak_check!(
        exist<> (for<'a, 'b> Equal<'a, 'b>)
    );
}

fn main() {}
