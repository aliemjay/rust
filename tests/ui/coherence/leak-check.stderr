error[E0119]: conflicting implementations of trait `CoherenceTrait` for type `test1::Unique`
  --> $DIR/leak-check.rs:27:9
   |
LL |   impl<T> CoherenceTrait for T {}
   |   ---------------------------- first implementation here
...
LL |           impl<$($lt),*> $crate::CoherenceTrait for Unique<$($lt),*>
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `test1::Unique`
...
LL | /     assert_false_by_leak_check!(
LL | |         exist<> (for<'a, 'b> Outlives<'a, 'b>)
LL | |     );
   | |_____- in this macro invocation
   |
   = note: this error originates in the macro `assert_false_by_leak_check` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `CoherenceTrait` for type `test3::Unique`
  --> $DIR/leak-check.rs:27:9
   |
LL |   impl<T> CoherenceTrait for T {}
   |   ---------------------------- first implementation here
...
LL |           impl<$($lt),*> $crate::CoherenceTrait for Unique<$($lt),*>
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `test3::Unique`
...
LL | /     assert_false_by_leak_check!(
LL | |         exist<> (for<'a> OutlivesPlaceholder<'a>)
LL | |     );
   | |_____- in this macro invocation
   |
   = note: this error originates in the macro `assert_false_by_leak_check` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0119]: conflicting implementations of trait `CoherenceTrait` for type `test6::Unique`
  --> $DIR/leak-check.rs:27:9
   |
LL |   impl<T> CoherenceTrait for T {}
   |   ---------------------------- first implementation here
...
LL |           impl<$($lt),*> $crate::CoherenceTrait for Unique<$($lt),*>
   |           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `test6::Unique`
...
LL | /     assert_false_by_leak_check!(
LL | |         exist<> (for<'a, 'b> Equal<'a, 'b>)
LL | |     );
   | |_____- in this macro invocation
   |
   = note: this error originates in the macro `assert_false_by_leak_check` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0119`.
